src/Symfony/Component/Mailer/Transport/Smtp/Stream/ProcessStream.php

```diff
@@ -0,0 +1,67 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Mailer\Transport\Smtp\Stream;
+
+use Symfony\Component\Mailer\Exception\TransportException;
+
+/**
+ * A stream supporting local processes.
+ *
+ * @author Fabien Potencier <fabien@symfony.com>
+ * @author Chris Corbyn
+ *
+ * @internal
+ *
+ * @experimental in 4.3
+ */
+final class ProcessStream extends AbstractStream
+{
+    private $command;
+
+    public function setCommand(string $command)
+    {
+        $this->command = $command;
+    }
+
+    public function initialize(): void
+    {
+        $descriptorSpec = [
+            0 => ['pipe', 'r'],
+            1 => ['pipe', 'w'],
+            2 => ['pipe', 'w'],
+        ];
+        $pipes = [];
+        $this->stream = proc_open($this->command, $descriptorSpec, $pipes);
+        stream_set_blocking($pipes[2], false);
+        if ($err = stream_get_contents($pipes[2])) {
+            throw new TransportException(sprintf('Process could not be started: %s.', $err));
```

Comment:
Personally I don't see the added value of using `sprintf`here, string concat is faster

Reply:
that's part of our coding standards (and such micro-optimization is not even worth it for the exception case, as it won't happen in a loop).